### ISSUE TITLE
Add bayesian skill rating

### ISSUE THREAD

**Issue by ETLegacyTeam** [etlegacy/etlegacy#403]
Author: @rmarquis (Spyhawk)
Date: 2018-12-23
Redmine Issue: 403, https://dev.etlegacy.com/issues/403

---

The idea is simple: Implement a new, proved metric that can be used to compare skill of players over time.

The existing XPs count measures player's score based on of specific action (kills, construction, revive, ..) but is far from evaluating all aspects of the 'skill' of a player. Also, XPs are useful to control the various skill levels, but don't give any indication of the real global performance - they basically only give a hint about the time the player has played on a specific server (xpsave).

A 'skill rating' could be defined as a metric measuring "all the parameters of a player that help his team to win" and the only way to do that is by a probabilistic, Bayesian approach. It simply should look at the global output of a game instead of the various micro-parameters (accuracy, number of kill, ... you-name-it).

ETPub implements such a metric with its "Player Rating", giving a normalized skill score. The TrueSkill system is a more modern algorythm that has been developed by MicroSoft for its XBox matching service. It has the advantage over the ETPub PR that it starts very low and increases over time (like XPs), before stabilizing when the skill rating is accurate (like the ETPub PR).

The algorythm is well documented (see Mooserware below) and several existing implementation can be found on the Internet, including this "Lua implementation":https://bitbucket.org/visionik/forged-alliance-forever-lua.

Also, we could use that Skill Rating to develop several nice feature later (showing probability of winning a map, team balancing, or even match making with players of similar skills, ... )

* http://en.wikipedia.org/wiki/TrueSkill
* http://www.moserware.com/2010/03/computing-your-skill.html

*Update:*

More detail and analysis is available on the wiki here "Bayesian skill rating":http://dev.etlegacy.com/projects/etlegacy/wiki/Bayesian_skill_rating

---

Related issue(s): #181
Child issue(s): #1139

---

Related commit(s):
a6dc54d6b9db4a44a9a6d6d8c780edf1d2f9af9a - misc: added optional FEATURE_RATING, refs 403
 0704e00bd0a736357ab9eebb714e81bad60336cc - misc: updated easybuild.sh, refs 403
 41a8f57c6ee8a936fee3957c0415edc69a409864 - game: added initial skill rating code, refs 403
 37037b62e7c89e9fb9f207eb27890733340f46e5 - game: added skill rating partial play, refs 403
 e2764f28624381a971b17768e0fe62483de48fbd - mod: added skill rating in scoreboard, refs 403
 f5e84f58776a726e280ebf2e7565f37eb4a967f7 - mod: added skill rating in stats window, refs 403
 b218a576e10e2dfc22a57d2e6a39806ccf3e429f - mod: added skill rating in debriefing window, refs..
 d11b073abf99baad6d0d01cf0088cdd509b9e561 - mod: added skill rating in console stats dump, ref..
 de4979d664d10b27e78a7432d640bc04fafe7355 - game: don't update skill rating in stopwatch and L..
 d89a352e758b1687ba30580ddb59446a462312b0 - mod: show rating scoreboard only if rating is enab..
 226ac1e1b33dbde2293e6a16d930af538367ff0c - mod: added win probability, refs 403
 f837d448360efea242dd53679517004735d210bd - mod: send only complete rating to clients, refs 4..
 ab09f9835347d5d57bb9ee00e6b860b954f66dd8 - mod: send delta rating to clients, refs 403
 8c155ab972429afd8ac090ba745cbe86bdf196e5 - game: extended rating algorithm with map parameter..
 903247dd77a827daf34b7af00cf5e2e9b6ab0cd4 - game: log match result, refs 403
 f07e062012c3274c0eccc547acf402db79cc89fe - game: removed debug stats, refs 403
 b79909ed35ca8a7b8aee61826caa5fd758a452ae - game: fix overwriting map vote values in CS_LEGACY..
 e8657df8c6ddcafb0fee41b3d4372b3a9be6f406 - mod: make g_skillRating latched. Don't execute rat..
 1005d01b2547cbe4af7a725531408c8e86a8a1af - mod: do not execute rating code when disabled, ref..
 46f4d36dd3e3f7bf60ed3d1f4c2053e14b0bfcfb - game: disable g_skillrating by default, refs 403
 c9118341e8e0b60614b0d484f24f6325150b8801 - cmake: enabled FEATURE_RATING by default, refs 40..
 bd8fc227a96892e34e830c2d13c09d8ceb5c6ec2 - game: fixed variable initialization, refs 403
 4ddfaa84ec797d8557dd7869f95d0eed89e3d5fa - cgame: removed useless code, refs 403
 40d67460ee730f24648b7b6605e3e0f4a3e86932 - cgame: fixed unintialized pointer read, refs 403
 86d98cd60e262ad0d8cb306bce2a068f04edf504 - game: added no division by zero check, refs 403..
 3b57ec90204dd2509d917d65ed69367cac2cfdaf - misc: enable FEATURE_RATING in easybuild.sh too, r..
 4b6ffab15baafacf9879497125150e3f023f446c - game: cleanup, refs 403
 83b26ab56a484036b2c04ab604d87c275e543011 - cgame: stats info label alignment, refs 403
 aa8ba49f1222f0cec01ee11ebffc6ae6c3afbbf3 - game: added missing map win bias, refs 403
 937d8a439606cc5569aad23847cc02ed89b15307 - cgame: display map prob on scoreboard, refs 403
 4fba92127fd44e03d047bd19cdeee07a97adc595 - mod: do not display negative rating and cleanup, r..
 506bf1d329381617e82174f7c9387ddc1e714e23 - cgame: fixed rating in stats window not updating w..
 b5ffb5b441ea644a1dde77dfdc01cccb5674867d - cgame: hide skill rating in debriefing when disabl..
 75e3180449f9896f6acc3df9ee09f685b7efa572 - cgame: fixed debriefing stats not displaying when ..
 845e23302558bbf73a44f821c11aa8cd79616e6e - mod: do not send any rating data when disabled, re..
 cba8e69743ebc5ce7fc7460f0a4f440ca2688f36 - mod: do not send delta rating in client userinfo, ..
 0416a7d7df23f5b130354605db0515c992c1f7f7 - mod: entirely removed rating from user configstrin..
 7ad92968567acf2b1c8b2061e7c40b34843ed47e - cgame: fix logical dead code (Coverity) + fix some..
 a19f62ea65ecc378d87bf696f7f332261cb315c6 - cgame: clean up + uncrustify
 9d97cf944ac67fd592b04b217b307b1f8ec5634c - game: fixed total play time, refs 403
 9d193942b83204c9adb47ab2a6e6cf1d9223a979 - mod: use highest ranking as first award, refs 403..
 52ee97dcc92c673ac6b55588e900dcfe990e1d35 - misc: added g_skillRating in default config file, ..
 efdfbdd1bcaedde1fc8183b10e93f17e69ba7afb - game: do not display meaningless total rating stat..
 29a636ced5a3cb9f9c3e2c1dd1a952c8b75dfa73 - game: fixed individual rating update with partial ..
 8681cbf454d5b33fb43c750ef5b219160735d3b4 - misc: enable FEATURE_DBMS if FEATURE_RATING is ena..
 d69d1ae1f627585ca7af9075154ca00f8ee43596 - game: use engine database for skill rating, refs ..
 8f3d76a21a23dcc7c11217d3f42aebd22e7c9219 - game: added map rating bias and continuity correct..
 ff73c3e6bb6a9ad7b78268e17e89fa70de8a9a38 - game: set RO flag on Lua rating variables, refs 4..
 b1590142e4fdf00cff40140c4c2537b025538c0a - game: set PRAGMA sync off with db_mode 2, refs 40..
 e56cbaae28f630e7f2e409b579e301efa765b1c6 - game: added shuffle by rating cmds, refs 403
 03c7fff92f541a5d132a07ba35e7c1ed22c052a5 - ui: added shuffle by rating menu options, refs 40..
 4a6029acad37a2e62ec4cbb4d32e0d838139b3fb - mod: added highest rated player award, refs 403 ..
 e26c6463e4e64022448cb96d51601385f99f8596 - game: set RO flag for time related Lua fields, ref..
 6a497024a590ec09fa4ea80b002871383bfda836 - game: fixed logically dead code, refs 403
 0b9bab5bf86c06d427088f920b8928d4d2daabcc - mod: display endgame stats float values with two d..
 f908a1c5073bcd2fae80bb0cfadbc339f03ec001 - game: skill rating send function cleanup, refs 40..
 c11f4030954147ceba939184b289172a58288f17 - game: win prob takes disconnected players into acc..
 f73bb19294e2e5515049d4307a0218788f388e77 - game: fixed missing guid assignment and small opti..
 240a61e9326209286e8f835c103a4304070c0b00 - game: code cleanup, refs 403
 1eb2ab85b8f4c71244585b5b1ae0a2b671495536 - game: do not display rating award if not positive,..
 69b25451db850053189026529d6a83c62825cfa4 - game: tied ranks to rating when enabled, refs 403
 27ab81371e39e92854f7915bb689a2d6fbd6e5bf - game: enabled skill rating by default, refs 403
 1bb1f37e9f3afc6dd5be22d06815ed1d2e472fcc - game: sanitize rank level and cleanup, refs 403
 777f46b25681cc20a63272c842bfa1a5c4457850 - game: capped display of max rating, refs 403
 a5ca3b40a35fe579027ab4a6c245282cb65c4158 - game: capped rating also in stats window and award..
 48802e570967a659e98de68c0bf00dc70166daff - game: update ranks at intermission, refs 403
 40559e74c7bebcd5f2cf7d3e063c6bb55bf1a3d0 - game: use map bias for slightly better rating shuf..
 7eb991b0056cab33e6517d6b7cc9c080a5a811f1 - game: fix compiling without FEATURE_RATING defined
 5fd94e727e8ec42145f0fc891828714abe9f9bd6 - Merge pull request 279 from zturtleman/skill-rati..
 b334986092424a2d86f95826ca9351646277ec48 - cgame: enabled SR scoreboard by default, refs 403..
 10dcedc91d346221cec19c051f0c466153455e7d - cgame: adjusted map bias display on SR scoreboard,..
 19420cb62db4af69c7ecd1e42be88e6df76f91c4 - cgame: display scoreboard type in infotext area, r..
 e23d262244a8b962084b350cad2e923192b92607 - cgame: fixed help string overlapping chat area, re..
 a282c2d3542b15cd29fba717eac35357ada08351 - cgame: removed scoreboard string to avoid overlapp..
 38999eb088f5b865ff4433598baf26de925459a6 - cgame: display scorboard mode in help text, refs ..
 711edfc747dbcae44f0427d6bee79b04acc7ef12 - cgame: hide scoreboard help text in intermission, ..
 4b6efa98eb34da65a00119cc83a11372c2158a2d - game: ensure SR db is deinit only once in intermis..
 2f2420213a77720212961fea76bf181da6e4d754 - cgame: fixed default scoreboard with g_SkillRating..
 15214acecbf76a4521230c36a63727b7cc97f770 - cgame: disabled SR scoreboard in SW and LMS mode, ..
 10068c8061bf7ad82817617835b9cc9cd5b81de4 - mod: only send delta rating in intermission, refs ..
 cfe54ca425eadba162e79272914a02492dded38a - mod: fix old demo playback, fixes 1320* refacto..
 8e610c3ebccbb2407adc76ba0a2dea616930abfc - game: minor skill rating code cleanup, refs 403
 8716200a4c801c07c919feffb1f91ec692817507 - game: added rating time init failsafe and cleanup,..
 8fa0ad26dede7e5375a024d967ff704175a9aef6 - game: fixed rating played time wrongly reset, refs..
 e52d73a27e8384bee3c0bac9e87f651933d99247 - game: fixed SR database not always available when ..
 164876b4f42a6cf61ab992eaf6344c6e28ed919c - mod: merged shuffle XP and SR commands, refs 403 ..
 790f8fc8a10e4632d4cc23f04c33168d4eceb4b3 - game: skill rating minor code cleanup, refs 403
 1ed90b03a738a13110b15160121b4ffeac8a09aa - game: remove min/max cap on displayed skill rating..
 96b083da1a363681e0f1519c0f0c481d7552d71e - game: added map bias display in statsdump, refs 4..
 d73285c5679450c17eb4356d7548d9a911af1c8a - game: added partial GUID in statsdump, refs 403
 14687e71fb70e19e3883a074131f5026aa4105b4 - game: log skill rating data, refs 403
 ccdfc12054d1cab0ef5d243cdc9be2e3c7bc5c5f - cgame: enabled statsdump autoaction by default, re..
 08bad8191287beccf7e782ff8175c90e0ca030af - game: fixed SR time data retrieval after reconnect..
 b6c3879c28e650701fb62aae142baa73415171de - game: fixed /statsdump not always displayed at end..
 f940824b756b8aa46de5e80d868e139aecf6f0a1 - game: minor macro fixup and cleanup, refs 403
 a5ef8f61955f016e3cc7ff442079acb20b9bbf15 - game: fixed partial GUID strings to avoid crash in..
 6c4cc0f5ae06e62ec1a0a7185d473edd7452acd0 - game: fixed guid array length in /players and end ..
 6600d972739b37aecf7ba137d3e97db36ce896e9 - game: disabled SR calculation for stopwatch and lm..
 a55a56c70c090941c57b7400b2288ab7f5ffee80 - game: minor skill rating code cleanup, refs 403
 ee708bff4f449299566252efcf927264204cd06c - game: decoupled game db init functions from SR cod..
 76ff0f305b36d607842a5ad78732c79113a5c423 - game: fixed database define, refs 403
 0cf54c90e854254635b15548dc2237c0f3ed02ac - game: use a better rank levels distribution, refs ..
 862601ed2f20a5f4f6509b86e6bc1d26915898f8 - game: log and display last win probability in stat..
 2bee354bbd43ffe0ea0e3f5d653482ba9298a7f0 - game: fixed win prob printing happens after statsd..
 00f6cf428b4553a6b90e123dadb71bdd057b7dbd - game: hide win prob in statsdump when SR is disabl..
 57fd7af9c16175502f0989262001ee118a1f36d8 - mod: hide gametype dependent stats in info window,..

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2015-09-02T19:52:28Z

---

Added link to some documentation in the wiki here: "Bayesian skill rating":http://dev.etlegacy.com/projects/etlegacy/wiki/Bayesian_skill_rating

**Comment by ETLegacyTeam** [comment]
From: @adanski (Dragonji)
Date: 2015-09-03T14:05:17Z

---

I'd say it would be nice to have it in the mod itself so it's easy to integrate it to current stats that are already in the game. It can't be nicely integrated in this aspect if it's a Lua script.

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2015-09-03T14:26:23Z

---

Dragonji wrote:
> I'd say it would be nice to have it in the mod itself so it's easy to integrate it to current stats that are already in the game. It can't be nicely integrated in this aspect if it's a Lua script.

You are right, and that is actually the way I'm implementing it. The algorithm itself and integration in the mod are mostly done, but I'm still wondering if the loading/writing of data itself would be better done in Lua.

Can't change the category of the ticket though, so it's stuck to "Lua scripts" for the time being.

**Comment by ETLegacyTeam** [comment]
From: @harleking (Harlekin)
Date: 2015-11-27T19:13:39Z

---

Would it be possible to let the server adjust some settings for a team based on team rating? Eg let them spawn a little faster, give them extra ammo on spawn, faster recharge of ammo/health cabinets, longer spawnprotection or some kind of other support for a very underpowered team.

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2015-11-27T19:32:57Z

---

Harlekin wrote:
> Would it be possible to let the server adjust some settings for a team based on team rating? Eg let them spawn a little faster, give them extra ammo on spawn, faster recharge of ammo/health cabinets, longer spawnprotection or some kind of other support for a very underpowered team.

Anything is possible, but as a first step we'll include only informative rating. Note that rating takes strength of teams into account, so it's actually "fair" to lose against a stronger team. But if you beat a stronger team, your rating would be increased much more than usual.

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2015-12-23T16:11:31Z

---

The base TrueSkill algorithm has been implemented (using session to keep values for now), and I extended it for taking map and time parameters.
However, to use this extension I'll need to collect data about played maps.

Data collected would be:
* number of time the map was played,
* number of time each side won the map,
* -average time before a team won the map- (not really necessary)

Here are a few possibilities:

1/ use a database through LuaSQL or the new engine DB feature;
2/ extend the existing mapvoteinfo.txt file system, and collect data in other gametypes (maybe apart from stopwatch);
3/ use a database system and move the existing mapvoteinfo data collect to a common database.

Option 3/ might be the cleanest, but 2/ will also work.
Opinion/suggestion?

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2016-08-30T11:41:54Z

---

{TODO}
* -Add missing Lua code (maybe as a WolfAdmin plugin)- (code is moved to C)
* -enable FEATURE_RATING- (done)
* -disable g_skillrating by default- (done)

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2016-09-02T14:53:57Z

---

I'll rework the existing mapvoteinfo data to use a common database system with the map parameter extension.

Edit: After some thought, the code has been modified so map stats management can be done entirely on the Lua side (using CS_LEGACYINFO). This should stay separate from the mapvote data.

Only the accompanying Lua script is missing now.

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2016-09-30T18:43:13Z

---

The Lua script has been partially written. It works and takes connected players at the end of the match to compute the rating update. Also, players that disconnected and reconnected later in the same match are correctly handled. However, to take the performance of all disconnected players into account, we need to loop over all players that played in the match, and not only level.numConnectedClients.

This require the use of an intermediary DB. I am unsure about the best way to implement this, as the C code needs to be adjusted but the DB stuff is for now entirely handled on the Lua side.

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2017-01-29T13:03:07Z

---

I'm hitting a wall with that DB-Lua stuff. I now believe the best way to do it is to implement directly in C, taking advantage of the engine DB.
It also makes sense since the rating system is deeply integrated in the mod code already.

Special commands can still be done in Lua, in the form of a WolfAdmin plugin or independent script.

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2017-02-23T10:09:43Z

---

{TODO}
* -Put Highest ranking officer award on top-
* -Add variable in config file-
* -Tie ranks to rating instead of XPs (see G_CalcRank() and G_UpgradeSkill(), might require explicit rank update, adjust the big PM message somehow)-
* -track number of matches played in DB- (not useful for rating)
* -use ribbon (like etnam)- (not useful for rating)
* -remove meaningless tot_rating in G_printMatchInfo()-
* -enable g_skillrating by default-
* -add new award for best rated player-
* -add shuffleteamsrating, shuffleteamsrating_norestart votes-
* -add info commands such as /rating, /allrating- (should be done in Lua)

WIP: move from Lua to C code: done!
* -DB Init in g_main-
* -DB DeInit in g_main.c-
* -DB Deinit in g_vote.c (map, campaign, nextmap, maprestart, mapreset)-
* -check users, match, map tables-
* -empty match table at init-
* -disable rating if failure-
* -ClientConnect -> retrieve rating from match or users tables, or create new one-
* -ClientDisconnect -> write in match table-
* -End of game G_LogExit(): adjust G_UpdateSkillRating for use of match table-
* -Add created and last updated time tracking-
* -Fix edge case player connects/disconnect in intermission-
* -Fix edge case player connects/disconnect in warmup-
* -Fix bug where rating from rating_users table in warmup isn't taken over playing session (use empty rating_match instead)-
* -Fix disconnected players aren't taken into account in scoreboard real time win probability-
* -Retrieve Map prob, Save map prob-
* -Display updated map rating in intermission scoreboard with g_skillRating 2-
* -Fix now broken delta rating-
* -Check log print G_Printf vs G_LogPrintf-
* -Proper headers doc-
* -Optimize speed with PRAGMA or single INSERT transaction for db_mode 2-
* -Set RO flag on Lua rating variables-
 
{FIXME}
* db_mode 1 with memory mode (see #1010)

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2017-03-11T01:27:29Z

---

Some results from the DB data after 1 week of data collection (initial mu: 25, initial sigma: 8.33, rating = mu - 3*sigma = 0):

* 73 entries, from which 18 are bots (55 humans players)
* best/worst rated player
** mu = 42.3, sigma = 6.17 (rating: 23.75). 
** mu = 20.6, sigma = 7.89 (rating: -3.10)
* human players with smallest sigma:
** mu = 19.3, sigma = 6 .15 (rating: 0.84)
** mu = 42.3, sigma = 6.17 (rating: 23.75)

This shows that ranks based on rating and distributed in a purely linear manner (0-5: Private, .., 45-50: General) are indeed viable. With the very limited number of players and initial bots ratings, it would requires about 3 weeks for the rating to stabilize (lower variance). The more players with higher rating (low uncertainty), the more the system can rate players quickly. 

However, using a more log-like distribution of ranks (lower ranks requiring a lower rating, f.e. 0-2: Private, 2-5: Private Fist Class, ..., 40-50: General) might increase diversity of ranks on a server and also motivate new players to stay on the same server.

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2017-03-12T13:42:56Z

---

Ranks are now tied to rating in a linear manner when rating is enabled. Otherwise, behavior is still the same with ranks tied to skill levels and XPs.

We lose the Ranks popup as the ratings and thus ranks aren't updated in-game but right before intermission, but I don't think it is worth adding complexity just to show these popup at the start of next map.

Note: An equivalent implementation to rank popup is tracked in #1139.

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2018-08-27T20:34:00Z

---

For the record: An "updated version of the TrueSkill algorithm":https://www.microsoft.com/en-us/research/uploads/prod/2018/03/trueskill2.pdf has been available for a few months:

Briefly, the classic TrueSkill model makes the following assumptions:
1. Each player has a latent skill value that represents their expected contribution to a team.
A player’s performance in a game is a noisy sample of their skill.
2. The performance of a team is the weighted sum of the performances of its players, where
the weight is the fraction of time the player spent on the team.
3. If a team’s performance is greater than the other team by a certain margin, the team wins.
Otherwise, the game is a draw. When learning skills from data, only the team win/loss
information is used (not scores).
4. Player skills evolve over time according to a random walk. An increase or decrease in skill
is assumed equally likely.
5. A player’s skill in a game mode is assumed independent of their skill in all other modes.

The TrueSkill2 model modifies the classic model in the following ways:
*1. A player’s latent skill is inferred from their individual statistics such as kill and death
counts, in addition to team win/loss.*
2. When a player quits or drops out in the middle of a game, it is treated as a surrender and
their skill is updated as if they lost a game (regardless of actual outcome).
3. A player’s skill in a game mode is assumed statistically correlated with their skill in other
modes, so that when a player starts a new mode, their skill rating from other modes is
borrowed.
4. The random walk of player skill is assumed biased toward increasing skill, with larger bias
during the first matches a player plays in a game mode.
5. When a player is part of a squad, their performance is assumed to be better than normal.

There is no existing implementation as of yet (only the paper). Might worth a look sometimes in the future.

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2019-07-12T10:54:34Z

---

{TODO} -Check for robustness of the db code.- (done, didn't found any issue)

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2019-11-28T18:22:40Z

---

{TODO}
* -check sess.time* values when player disconnect then reconnect on the same match.- (done, fixed data not retrieved and refactored db code)
